### PR TITLE
UBERF-8657 API Client package

### PR DESCRIPTION
**Pull request by aonnikov**

### CODE REVIEW

File: packages/api-client/src/account.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      params: [user, password, workspace]
+    })
+  })
+
+  const result = await response.json()
+  return result.result?.token
+}
+
+/** @public */
+export async function selectWorkspace (
```
Comment (haiodo) [inline packages/api-client/src/account.ts:46]: let's not expose close this method.

File: packages/api-client/src/client.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+  const { endpoint, token } = await getWorkspaceToken(url, options, config)
+  return await createClient(endpoint, token, options)
+}
+
+/** @public */
+export async function createClient (endpoint: string, token: string, options: ConnectSocketOptions): Promise<APIClient> {
+  addLocation(clientId, () => import(/* webpackChunkName: "client" */ '@hcengineering/client-resources'))
+
+  setMetadata(client.metadata.ClientSocketFactory, options.socketFactory)
```
Comment (haiodo) [inline packages/api-client/src/client.ts]: let's move all this 3 as parameters to client factory.

File: packages/api-client/src/socket/node.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import { type ClientSocket, type ClientSocketFactory } from '@hcengineering/client'
+import WebSocket from 'ws'
```
Comment (haiodo) [inline packages/api-client/src/socket/node.ts]: please move to factory as require('ws') and make dependency optional.

File: packages/api-client/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  workspace: string
+}
+
+/** @public */
+export type AuthOptions = PasswordAuthOptions | TokenAuthOptions
+
+/** @public */
+export interface ConnectSocketOptions {
+  socketFactory?: ClientSocketFactory
+  useBinaryProtocol?: boolean
```
Comment (haiodo) [inline packages/api-client/src/types.ts]: this should not be exposed

Comment (aonnikov) [reply packages/api-client/src/types.ts]: Only `useBinaryProtocol` or both `useXXX`?

File: packages/api-client/src/account.ts (hunk 1 above)
Comment (aonnikov) [reply packages/api-client/src/account.ts:46]: this `account` module is internal and not supposed to be exposed to users. Is that ok or you want to do something else?

File: packages/api-client/src/client.ts (hunk 2 above)
Comment (aonnikov) [reply packages/api-client/src/client.ts]: done

File: packages/api-client/src/socket/node.ts (hunk 3 above)
Comment (aonnikov) [reply packages/api-client/src/socket/node.ts]: done

File: packages/api-client/src/types.ts (hunk 4 above)
Comment (haiodo) [reply packages/api-client/src/types.ts]: This is internal so, I think we do not need to expose them both.

File: packages/api-client/src/account.ts (hunk 1 above)
Comment (haiodo) [reply packages/api-client/src/account.ts:46]: internal is fine.

File: packages/api-client/src/types.ts (hunk 4 above)
Comment (aonnikov) [reply packages/api-client/src/types.ts]: removed both

**Review by haiodo — APPROVED** [review]

**Pull request merged**